drivers/dose/dose.c

```diff
@@ -134,49 +160,79 @@ static dose_signal_t state_transit_idle(dose_t *ctx, dose_signal_t signal)
-        }
-        else if (!esc && ctx->uart_octet == DOSE_OCTET_END) {
-            SETBIT(ctx->flags, DOSE_FLAG_END_RECEIVED);
-            rc = DOSE_SIGNAL_END;
-        }
-        else {
-            if (esc) {
-                CLRBIT(ctx->flags, DOSE_FLAG_ESC_RECEIVED);
+        const dose_header_t *header = (dose_header_t *)ctx->header_recv_buf;
+        bool sof = ctx->flags & DOSE_FLAG_SOF_RECEIVED;
+        bool have_len = ctx->header_buf_ptr == DOSE_FRAME_HEADER_LEN;
+
+        /* no start of frame marker found yet */
+        if (!sof) {
+            assert(!have_len);
+            switch (ctx->uart_octet) {
+            case DOSE_OCTET_PREAMBLE:
+                /* there might be multiple preamble octets, ignore them */
+                return DOSE_SIGNAL_NONE;
+            case DOSE_OCTET_SOF:
+                /* start of frame marker found, begin frame processing */
+                SETBIT(ctx->flags, DOSE_FLAG_SOF_RECEIVED);
+                _set_timeout_bytes(ctx, 4 * DOSE_FRAME_HEADER_LEN);
+                break;
+            default:
+                /* invalid octet, discard frame */
+                DEBUG_PUTS("dose: bad octet");
+                return DOSE_SIGNAL_END;
             }
+        }
+
+        if (have_len) {
             /* Since the dirty flag is set after the RECV state is left,
              * it indicates that the receive buffer contains unprocessed data
              * from a previously received frame. Thus, we just ignore new data. */
-            if (!(ctx->flags & DOSE_FLAG_RECV_BUF_DIRTY)
-                && ctx->recv_buf_ptr < DOSE_FRAME_LEN) {
-                ctx->recv_buf[ctx->recv_buf_ptr++] = ctx->uart_octet;
+            if (ctx->flags & DOSE_FLAG_RECV_BUF_DIRTY) {
+                DEBUG_PUTS("dose: rx buffer locked");
+                return DOSE_SIGNAL_NONE;
             }
-        }
-    }
 
-    if (rc == DOSE_SIGNAL_NONE) {
-        /* No signal is returned. We stay in the RECV state. */
-        xtimer_set(&ctx->timeout, ctx->timeout_base);
+            ctx->recv_buf[ctx->recv_buf_ptr++] = ctx->uart_octet;
+
+            if (ctx->recv_buf_ptr == header->frame_len) {
+                SETBIT(ctx->flags, DOSE_FLAG_END_RECEIVED);
+                return DOSE_SIGNAL_END;
+            }
+        } else {
+
+            ctx->header_recv_buf[ctx->header_buf_ptr++] = ctx->uart_octet;
+
+            if (ctx->header_buf_ptr == DOSE_FRAME_HEADER_LEN) {
+                /* invalid parity */
+                if ((ctx->header_recv_buf[1] ^
+                     ctx->header_recv_buf[2] ^
+                     ctx->header_recv_buf[3]) || header->frame_len > DOSE_FRAME_LEN) {
+                    clear_recv_header(ctx);
+                    DEBUG_PUTS("dose: invalid len");
+                    return DOSE_SIGNAL_END;
+                }
+                /* set timeout based on frame length, sender might be slow */
+                _set_timeout_bytes(ctx, 4 * header->frame_len);
```

Comment:
Here the wire-transfer-factor of 4 per byte re-appears. If you made some measurements to back this it would be good to put the factor into a define accompanied by some explanation to make conditions reproducible for future contributions. Maybe people would even like to have this grace factor configurable.

Reply:
The factor of 4 is just a good guess, I did not do measurements. I chose 4 because it converts into shifts easily and the delay between two bytes might be larger than a single byte.

I'll do some measurements on the oscilloscope to see how big the delay between bytes can be on an unoptimized sender.